File: src/services/logging.ts
import { invoke } from '@tauri-apps/api/core';

/** Log a simple message to the app's log.txt file. */
export async function logMessage(message: string): Promise<void> {
  try {
    await invoke('log_message', { message });
  } catch (e) {
    console.error('[Logger] log_message failed:', e);
  }
}

/** Log a tagged debug entry with JSON payload. */
export async function logDebug(tag: string, payload: unknown): Promise<void> {
  try {
    const jsonStr =
      typeof payload === 'string' ? payload : JSON.stringify(payload);
    await invoke('log_debug', { tag, payload: jsonStr });
  } catch (e) {
    console.error('[Logger] log_debug failed:', e);
  }
}

/** Log a tagged entry with a JSON-serializable object. */
export async function logJson(tag: string, payload: unknown): Promise<void> {
  try {
    await invoke('log_json', { tag, payload });
  } catch (e) {
    console.error('[Logger] log_json failed:', e);
  }
}

File: src-tauri/src/logging.rs
use std::fs::OpenOptions;
use std::io::Write;
use chrono::Local;

/// Append a timestamped message to the app's log.txt file.
#[tauri::command]
pub fn log_message(message: String) -> Result<(), String> {
    let log_path = std::path::PathBuf::from("log.txt");
    let timestamp = Local::now().format("%Y-%m-%d %H:%M:%S%.3f");
    let entry = format!("[{}] {}\n", timestamp, message);

    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(&log_path)
        .map_err(|e| format!("Failed to open log file: {}", e))?;

    file.write_all(entry.as_bytes())
        .map_err(|e| format!("Failed to write log: {}", e))?;

    Ok(())
}

/// Log a structured JSON entry for debugging IPC flows.
#[tauri::command]
pub fn log_debug(tag: String, payload: String) -> Result<(), String> {
    let log_path = std::path::PathBuf::from("log.txt");
    let timestamp = Local::now().format("%Y-%m-%d %H:%M:%S%.3f");
    let entry = format!("[{}] [{}] {}\n", timestamp, tag, payload);

    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(&log_path)
        .map_err(|e| format!("Failed to open log file: {}", e))?;

    file.write_all(entry.as_bytes())
        .map_err(|e| format!("Failed to write log: {}", e))?;

    Ok(())
}

/// Convenience: log a tag + JSON-serializable payload in one call.
#[tauri::command]
pub fn log_json(tag: String, payload: serde_json::Value) -> Result<(), String> {
    log_debug(tag, serde_json::to_string(&payload).unwrap_or_default())
}